core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -466,14 +479,14 @@ public void streamCreated(ServerStream stream, String methodName, Metadata heade
 
     private void streamCreatedInternal(
         final ServerStream stream, final String methodName, final Metadata headers, final Tag tag) {
-      final Executor wrappedExecutor;
+      final Executor defaultExecutor;
       // This is a performance optimization that avoids the synchronization and queuing overhead
       // that comes with SerializingExecutor.
       if (executor == directExecutor()) {
```

Comment:
It seems like we'll need to disable these two optimizations if an executor supplier is provided. I think the threading for optimizeForDirectExecutor requires it to be done inline before this method returns. It is safe to call it even if direct executor isn't used, but if/when migrating deframer gets enabled again getting it wrong hurts the non-direct case much more than it helps the direct case; that is to say, if we don't know which executor will be used, it is better to error on the side of not calling the method.

Reply:
But executor supplier is never used here when direct executor is enabled.  It looks optimizeForDirectExecutor just means we should use SerializeReentrantCallsDirectExecutor, and that would run tasks inline but still queued if already executing, so, it is still safe when we would have two contextRunnables, because listener call backs would still be queued at  SerializeReentrantCallsDirectExecutor.  
How does migrating deframer hurt the non-direct case?
Did you mean we should not allow them to co-exist? 
`executor==directExecutor and executor_supplier != null` validation error at server builder.
`executor==directExecutor and executor_supplier == null`: optimizeForDirectExecutor and use SerializeReentrantCallsDirectExecutor
`executor!=directExecutor and executor_supplier != null`: Just SerializingExecutor, no optimization.
`executor!=directExecutor and executor_supplier == null`: always use executor(I need to check and fix)

Reply:
> But executor supplier is never used here when direct executor is enabled.

But that is clearly not right/incomplete. The executor for lookups in the fallback registry may be direct executor and that does not seem to be cause to ignore a user's executor supplier. I was discussing how we'll go about mixing all this stuff together.

> It looks optimizeForDirectExecutor just means we should use SerializeReentrantCallsDirectExecutor

SerializeReentrantCallsDirectExecutor is different from optimizeForDirectExecutor. optimizeForDirectExecutor was intended to disable MigratingDeframer when using direct executor, although MigratingDeframer is disabled for all cases at the moment.

> How does migrating deframer hurt the non-direct case?

It _helps_ the non-direct case. However, it adds overhead that can hurt the direct case, so the code was disabling it when direct executor was used. I was saying the overhead it adds to direct is pretty small compared to how much it helps the non-direct case; we can simply disable this optimization (i.e., not call optimizeForDirectExecutor) if an executor supplier is provided.

> executor==directExecutor and executor_supplier != null validation error at server builder.

I see no reason to cause an error. The two conceptually can make sense. It is just a question of how many optimizations does it disable. I'm suggesting we'll treat this the same as `executor!=directExecutor and executor_supplier != null`.